sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/estimation/FilterEstimation.scala

```diff
@@ -0,0 +1,479 @@
+      literal: Literal,
+      update: Boolean)
+    : Double = {
+    if (!mutableColStats.contains(attrRef.exprId)) {
+      logDebug("[CBO] No statistics for " + attrRef)
+      return 1.0
+    }
+
+    /** Make sure that the Date/Timestamp literal is a valid one */
+    attrRef.dataType match {
+      case DateType =>
+        val dateLiteral = DateTimeUtils.stringToDate(literal.value.asInstanceOf[UTF8String])
+        if (dateLiteral.isEmpty) {
+          logDebug("[CBO] Date literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+      case TimestampType =>
+        val tsLiteral = DateTimeUtils.stringToTimestamp(literal.value.asInstanceOf[UTF8String])
+        if (tsLiteral.isEmpty) {
+          logDebug("[CBO] Timestamp literal is wrong, No statistics for " + attrRef)
+          return 1.0
+        }
+      case _ =>
+    }
+
+    op match {
+      case EqualTo(l, r) => evaluateEqualTo(op, attrRef, literal, update)
+      case _ =>
+        attrRef.dataType match {
+          case _: NumericType | DateType | TimestampType =>
+            evaluateBinaryForNumeric(op, attrRef, literal, update)
+          case StringType | BinaryType =>
+            /**
+             * TODO: It is difficult to support other binary comparisons for String/Binary
+             * type without min/max and advanced statistics like histogram.
+             */
+            logDebug("[CBO] No statistics for String/Binary type " + attrRef)
+            return 1.0
+        }
+    }
+  }
+
+  /**
+   * This method converts a numeric or Literal value of numeric type to a BigDecimal value.
+   * If isNumeric is true, then it is a numeric value.  Otherwise, it is a Literal value.
+   */
+  def numericLiteralToBigDecimal(
+       literal: Any,
+       dataType: DataType,
+       isNumeric: Boolean = false)
+    : BigDecimal = {
+    dataType match {
+      case _: IntegralType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[Long])
+        else BigDecimal(literal.asInstanceOf[Literal].value.asInstanceOf[Long])
+      case _: FractionalType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[Double])
+        else BigDecimal(literal.asInstanceOf[Literal].value.asInstanceOf[Double])
+      case DateType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[BigInt])
+        else {
+          val dateLiteral = DateTimeUtils.stringToDate(
+            literal.asInstanceOf[Literal].value.asInstanceOf[UTF8String])
+          BigDecimal(dateLiteral.asInstanceOf[BigInt])
+        }
+      case TimestampType =>
+        if (isNumeric) BigDecimal(literal.asInstanceOf[BigInt])
+        else {
+          val tsLiteral = DateTimeUtils.stringToTimestamp(
+            literal.asInstanceOf[Literal].value.asInstanceOf[UTF8String])
+          BigDecimal(tsLiteral.asInstanceOf[BigInt])
+        }
+    }
+  }
+
+  /** This method evaluates the equality predicate for all data types. */
+  def evaluateEqualTo(

```

Comment:
Yes, the return value is a doube value showing the percentage of rows meeting a given condition.  Also I will add comments for this method in JavaDoc style.